crates/bevy_asset/src/http_source.rs

```diff
@@ -0,0 +1,244 @@
+use crate::AssetApp;
+use bevy_app::App;
+use bevy_utils::ConditionalSendFuture;
+use std::path::{Path, PathBuf};
+
+/// Adds the `http` and `https` asset sources to the app.
+/// Any asset path that begins with `http` or `https` will be loaded from the web
+/// via `fetch`(wasm) or `ureq`(native).
+pub fn http_source_plugin(app: &mut App) {
+    app.register_asset_source(
+        "http",
+        AssetSource::build().with_reader(|| Box::new(HttpSourceAssetReader::Http)),
+    );
+    app.register_asset_source(
+        "https",
+        AssetSource::build().with_reader(|| Box::new(HttpSourceAssetReader::Https)),
+    );
+}
+
+/// Asset reader that treats paths as urls to load assets from.
+pub enum HttpSourceAssetReader {
+    /// Unencrypted connections.
+    Http,
+    /// Use TLS for setting up connections.
+    Https,
+}
+
+impl HttpSourceAssetReader {
+    fn make_uri(&self, path: &Path) -> PathBuf {
+        PathBuf::from(match self {
+            Self::Http => "http://",
+            Self::Https => "https://",
+        })
+        .join(path)
+    }
+
+    /// See [`crate::io::get_meta_path`]
+    fn make_meta_uri(&self, path: &Path) -> Option<PathBuf> {
+        let mut uri = self.make_uri(path);
+        let mut extension = path.extension()?.to_os_string();
+        extension.push(".meta");
+        uri.set_extension(extension);
+        Some(uri)
+    }
+}
+
+#[cfg(target_arch = "wasm32")]
+async fn get<'a>(path: PathBuf) -> Result<Box<dyn Reader>, AssetReaderError> {
+    use crate::io::wasm::HttpWasmAssetReader;
+
+    HttpWasmAssetReader::new("")
+        .fetch_bytes(path)
+        .await
+        .map(|r| Box::new(r) as Box<dyn Reader>)
+}
+
+#[cfg(not(target_arch = "wasm32"))]
+async fn get<'a>(path: PathBuf) -> Result<Box<dyn Reader>, AssetReaderError> {
+    use crate::io::VecReader;
+    use std::io;
+
+    let str_path = path.to_str().ok_or_else(|| {
+        AssetReaderError::Io(
+            io::Error::new(
+                io::ErrorKind::Other,
+                format!("non-utf8 path: {}", path.display()),
+            )
+            .into(),
+        )
+    })?;
+
+    #[cfg(feature = "http_source_cache")]
+    if let Some(data) = http_asset_cache::try_load_from_cache(str_path)? {
+        return Ok(Box::new(VecReader::new(data)));
+    }
+
+    match ureq::get(str_path).call() {
+        Ok(response) => {
+            let mut reader = response.into_reader();
+            let mut buffer = Vec::new();
+            reader.read_to_end(&mut buffer)?;
+
+            #[cfg(feature = "http_source_cache")]
+            http_asset_cache::save_to_cache(str_path, &buffer)?;
+
+            Ok(Box::new(VecReader::new(buffer)))
+        }
+        // ureq considers all >=400 status codes as errors
+        Err(ureq::Error::Status(code, _response)) => {
+            if code == 404 {
+                Err(AssetReaderError::NotFound(path))
+            } else {
+                Err(AssetReaderError::HttpError(code))
```

Comment:
Maybe helpful to add some more error context here by also capturing the path in these cases?

Reply:
Its a good idea that should probably be applied to all variants of `AssetReaderError`. The change is straightforward but touches about a dozen files and causes things like `PartialEq` to work differently, could be more cleanly done in a seperate pr?

Reply:
Agreed. You could cheat this by wrapping the top-most call in an IO error with the path in the description and the original error nested.